docs/spec/fastsync/scheduler.tla

```diff
@@ -0,0 +1,604 @@
+
+\* messages from scheduler
+OnScFinishedEv ==
+  /\ outEvent.type = "scFinishedEv"
+  /\ envRunning' = FALSE \* stop the env
+  /\ UNCHANGED inEvent
+
+NextEnv ==
+  IF ~envRunning THEN
+    UNCHANGED <<inEvent, envRunning>>
+  ELSE
+    \/ OnScFinishedEv
+    \/ OnGlobalTimeoutTicker
+    \/ OnAddPeerEv
+    \/ OnTrySchedule
+    \/ OnStatusResponseEv
+    \/ OnBlockResponseEv
+    \/ OnNoBlockResponseEv
+    \/ OnRemovePeerEv
+    \/ OnPcBlockProcessed
+    \/ OnPcBlockVerificationFailure
+
+(* ----------------------------------------------------------------------------------------------*)
+(* The system is the composition of the environment and the schedule                             *)
+(* ----------------------------------------------------------------------------------------------*)
+Init == turn = "environment" /\ InitEnv /\ InitSc
+
+FlipTurn == 
+  turn' = (
+    IF turn = "scheduler" THEN 
+      "environment" 
+    ELSE
+      "scheduler"
+  ) 
+
+\* scheduler and environment alternate their steps (synchronous composition introduces more states)
+Next ==
+/\ FlipTurn
+/\ IF turn = "scheduler" THEN
+     /\ NextSc
+     /\ inEvent' = noEvent
+     /\ UNCHANGED envRunning
+   ELSE
+     /\ NextEnv
+     /\ outEvent' = noEvent
+     /\ UNCHANGED <<scRunning, scheduler>>
+
+Spec == Init /\ [][Next]_vars /\ WF_turn(FlipTurn)
+
+(* ----------------------------------------------------------------------------------------------*)
+(* Invariants                                                                                    *)
+(* ----------------------------------------------------------------------------------------------*)
+TypeOK ==
+    /\ turn \in {"scheduler", "environment"}
+    /\ inEvent \in InEvents
+    /\ envRunning \in BOOLEAN
+    /\ outEvent \in OutEvents
+    /\ scheduler \in [
+         initHeight: Heights,
+         height: Heights \cup {ultimateHeight + 1},
+         peers: SUBSET PeerIDs,
+         peerHeights: [PeerIDs -> Heights \cup {None}],
+         peerStates: [PeerIDs -> PeerStates],
+         blocks: SUBSET Heights,
+         blockStates: [Heights -> BlockStates],
+         pendingBlocks: [Heights -> PeerIDs \cup {None}],
+         receivedBlocks: [Heights -> PeerIDs \cup {None}]
+       ]
+
+(* ----------------------------------------------------------------------------------------------*)
+(* Helpers for Properties                                                                        *)
+(* ----------------------------------------------------------------------------------------------*)
+NoFailuresAndTimeouts ==
+    /\ inEvent.type /= "bcRemovePeer"
+    /\ inEvent.type /= "bcNoBlockResponse"
+    /\ inEvent.type /= "pcBlockVerificationFailure"
+
+\* simulate good peer behavior using this formula. Useful to show termination in the presence of good peers.
+GoodResponse ==
+  \/ inEvent.type
+         \in {"bcAddNewPeer", "bcStatusResponse", "bcBlockResponse", "pcBlockProcessed", "rTrySchedule"}
```

Comment:
We should remove `rTrySchedule` in case we don't want to implement the corresponding handler.

Reply:
there is `handleTrySchedule` that uses it.